src/main/java/org/junit/runners/model/TestClass.java

```diff
@@ -255,14 +284,13 @@ public String getName() {
                  */
                 if (valueClass.isAssignableFrom(each.getReturnType())) {
                     Object fieldValue = each.invokeExplosively(test);
-                    results.add(valueClass.cast(fieldValue));
+                    consumer.accept(each, valueClass.cast(fieldValue));
```

Comment:
One thing I noticed in my attempt is that we will call `MethodRule` methods twice if the return value implements both `TestRule` and `MethodRule` (one of the values would be thrown away in `withRules()`. My `MethodValueCollector` class avoided this, because the collector subclass can filter out the values before the method value was called.

I'm not sure how much we should worry about that. Your approach is quite nice.